[TE] fix(rdma): pre-compute slice_count to avoid race condition

wwq2333 opened the pull request:
## Description

Previously, `task.slice_count` was incremented inside the slicing loop using `__sync_fetch_and_add()`. This caused a race condition under USE_EVENT_DRIVEN_COMPLETION (#1053): when some slices completed quickly, `completed_slice_count` could equal `slice_count` before all slices were created, causing premature task completion detection. Similarly, any other logic that needs to access `slice_count` could also be affected by this issue.

This commit:
- Pre-computes the total slice count before entering the slicing loop based on request.length, kBlockSize, and kFragmentSize (matching the existing merge logic)
- Removes the incremental __sync_fetch_and_add() call inside the loop
- Removes incorrect manual slice cleanup in the error path (TransferTask destructor already handles slice deallocation, manual cleanup caused double-free)

## Type of Change

* Types
  - [X] Bug fix
  - [ ] New feature
    - [ ] Transfer Engine
    - [ ] Mooncake Store
    - [ ] Mooncake EP
    - [ ] Integration
    - [ ] P2P Store
    - [ ] Python Wheel
  - [ ] Breaking change
  - [ ] CI/CD
  - [ ] Documentation update
  - [ ] Other

## How Has This Been Tested?

<!-- Please describe the tests you've run to verify your changes. -->

## Checklist

- [X] I have performed a self-review of my own code.
- [ ] I have updated the documentation.
- [ ] I have added tests to prove my changes are effective.

Commit:
[TE] fix(rdma): pre-compute slice_count to avoid race condition

Commit:
[TE] fix(rdma): simplify slice count calculation by merging redundant conditions

staryxchen reviewed (commented):
LGTM. Have you run tests to verify that this PR is worked?

staryxchen reviewed (approved):
LGTM. 
BTW, I accidentally closed this PR earlier. My apologies.

wwq2333 commented on mooncake-transfer-engine/src/transport/rdma_transport/rdma_transport.cpp:
```diff
@@ -323,6 +323,33 @@ Status RdmaTransport::submitTransferTask(
+            }
+            task.slice_count = total_slices;
+        }
+
```
Do you mean extracting a unified function that handles both calculating and allocating slices? For example:
```
// If slices_to_post is provided (non-null), allocate and populate slices.
// Otherwise, only compute task.slice_count (and task.total_bytes).
Status prepareSlices(
    TransferTask &task,
    std::unordered_map<RdmaContext *, std::vector<Slice *>> *slices_to_post = nullptr);
```

Commit:
[TE] fix(rdma): unify slice counting and allocation logic

Extract slice preparation into a single prepareTaskSlices() method that
handles both slice count calculation and slice allocation. This ensures
the slicing logic is maintained in exactly one place, preventing
inconsistencies when the slicing rules are modified.

The submitTransferTask() now executes in two phases:
- Phase 1: Pre-compute slice_count and total_bytes for all tasks
- Phase 2: Allocate slices and submit to workers

This fixes a race condition where completed slices could be counted
against an incomplete slice_count, causing premature task completion
detection.

Also introduce SliceAllocateConfig struct to avoid redundant
globalConfig() calls across task iterations.
